port/raspberrypi/rp2xxx/src/hal/dma.zig

```diff
@@ -102,13 +130,201 @@ pub const Channel = enum(u4) {
         regs.read_addr = read_addr;
         regs.write_addr = write_addr;
         regs.trans_count = count;
-        regs.ctrl_trig.modify(.{
-            .EN = @intFromBool(config.enable),
-            .DATA_SIZE = config.data_size,
-            .INCR_READ = @intFromBool(config.read_increment),
-            .INCR_WRITE = @intFromBool(config.write_increment),
-            .TREQ_SEL = config.dreq,
+        if (config.trigger) {
+            regs.ctrl_trig.modify(.{
+                .EN = @intFromBool(config.enable),
+                .DATA_SIZE = config.data_size,
+                .INCR_READ = @intFromBool(config.read_increment),
+                .INCR_WRITE = @intFromBool(config.write_increment),
+                .TREQ_SEL = config.dreq,
+            });
+        } else {
+            regs.al1_ctrl.modify(.{
+                .EN = @intFromBool(config.enable),
+                .DATA_SIZE = config.data_size,
+                .INCR_READ = @intFromBool(config.read_increment),
+                .INCR_WRITE = @intFromBool(config.write_increment),
+                .TREQ_SEL = config.dreq,
+            });
+        }
+    }
+
+    pub const SetupTransferConfig = struct {
+        trigger: bool = false,
+        enable: bool,
+    };
+
+    pub fn setup_transfer(
+        chan: Channel,
+        write: anytype,
+        read: anytype,
+        config: SetupTransferConfig,
+    ) !void {
+        const H = struct {
+            fn is_peripheral(Type: type) bool {
+                return Type == DmaReadTarget or Type == DmaWriteTarget;
+            }
+
+            fn validate_type(Type: type) void {
+                const Info = @typeInfo(Type);
+                switch (Info) {
+                    .@"struct" => {
+                        if (!is_peripheral(Type))
+                            @compileError("only perihperals and pointers are supported");
+                    },
+                    .pointer => {
+                        if (get_data_size(Type) == null)
+                            @compileError("only pointers/slices/arrays of u8/u16/u32 are supported");
+                    },
+                    else => @compileError(std.fmt.comptimePrint("unsupported type {}", .{Type})),
+                }
+            }
+
+            inline fn get_addr(value: anytype) u32 {
+                const Type = @TypeOf(value);
+                const Info = @typeInfo(Type);
+                switch (Info) {
+                    .@"struct" => {
+                        return value.addr;
+                    },
+                    .pointer => {
+                        return @intFromPtr(value);
+                    },
+                    else => comptime unreachable,
```

Comment:
```suggestion
                return switch (Info) {
                    .@"struct" => value.addr
                    .pointer => @intFromPtr(value);
                    else => comptime unreachable,
```
Also, not sure if you need the `comptime`. Anything with types is de-facto comptime, so you could remove that keyboard and @compileError